### PR TITLE
Implement fuel price module

### PR DESCRIPTION
**Pull request by Brod8362**
This implements the functional part of the fuel price module.
There's still work to do regarding formatting and such, but the functionality is there.

### CODE REVIEW

**Comment by tnttruong** [comment]
image
I'm getting a panic upon submitting fuel quote, reproducible by inputting fuel quote data and then pressing the submit button

**Comment by Brod8362** [comment]
Does the account you're using have profile information set?

**Comment by Brod8362** [comment]
This will occur if you don't have profile information - that will be fixed in a separate issue. Set profile information and see if you can still reproduce

**Comment by tnttruong** [comment]
Yes this works when profile information is formatted correctly, it was an error on my end

File: src/backend/src/routes.rs [hunk 1]
```diff
@@ -242,25 +255,30 @@ pub fn quote() -> Template {
...
 #[post("/page/quote", data="<form>")]
-pub async fn quote_submit(ctx: &UserContext, pool: &State<Pool<Sqlite>>, form: Form<QuoteData>) -> Template {
-    let submit_resp = database::store_quote(pool, &ctx.user_id, form.gallons_requested, &form.address, &form.delivery_date).await;
+pub async fn quote_submit(ctx: &UserContext, pool: &State<Pool<Sqlite>>, form: Form<PartialQuote>) -> Template {
+    //address should come from the profile
+    //TODO: expect here bad, but is temporary
+    let profile_maybe = database::get_profile(pool, &ctx.user_id).await.expect("failed to get user profile");
+    if profile_maybe.is_none() {
+        //TODO error
+    }
```
Comment (tnttruong) [inline src/backend/src/routes.rs:270]: TODO reminder here unless its unnecessary

File: src/backend/src/routes.rs [hunk 2]
```diff
@@ -187,6 +190,16 @@ pub struct ProfileData {
...
 }
 
+impl ProfileData {
+    pub fn is_in_state(&self) -> bool {
+        self.state == "TX"
+    }
+
+    pub fn one_line_address(&self) -> String {
+        String::from("TODO") //TODO: put michelle's logic here
+    }
```
Comment (tnttruong) [inline src/backend/src/routes.rs:200]: reminder on merge to do this

File: src/backend/src/routes.rs [hunk 3]
```diff
@@ -272,6 +290,28 @@ pub async fn quote_submit(ctx: &UserContext, pool: &State<Pool<Sqlite>>, form: F
...
 }
 
+#[post("/page/quote/preview", data="<form>")]
+pub async fn quote_preview(ctx: &UserContext, pool: &State<Pool<Sqlite>>, form: Form<PartialQuote>) -> Template {
+    //TODO: don't .expect, properly handle the error
+    //don't store the quote, just generate the pricing info
+    let profile_maybe = database::get_profile(&pool, &ctx.user_id).await.expect("failed to get profile data");
+    if profile_maybe.is_none() {
+        //TODO error here
+    }
```
Comment (tnttruong) [inline src/backend/src/routes.rs:300]: todo reminder here unless its unnecessary

**Review by tnttruong — DISMISSED** [review]
Runs perfectly for me, there's just some TODOs leftover in your code that should be handled upon merge though

**Comment by Brod8362** [comment]
The TODOs will be resolved in another PR after #50 is merged

File: src/backend/src/pricing.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    Ok(pricing_core(in_state, has_requested_previously, qty))
+}
+
+fn pricing_core(in_state: bool, has_requested: bool, qty: f64) -> (f64, f64) {
+    // 2% for in-state, 4% for out of state
+    let location_factor = if in_state { 0.02 } else { 0.04 };
+
+    // 0% if you've requested before, 1% if you haven't
+    let history_factor = if has_requested { 0.00 } else { 0.01 };
```
Comment (Mnguyener) [inline src/backend/src/pricing.rs:26]: Should we make a note for users to know about the history_factor?

File: src/backend/templates/fuel_quote_preview.html.tera [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<div class="form-sect">
+    <!-- this input is not actually used, just visual -->
+    <img src="/static/money_icon1.png" alt="money icon">
+    <label for="pergallon">$/gal:</label>
+    <input name="pergallon" value="${{per_gallon}}" disabled>
+    <!-- price is a noneditable numeric value calculated by pricing module-->
+</div>
+
+<div class="form-sect">
+    <!-- this input is not actually used, its just visual -->
```
Comment (Mnguyener) [inline src/backend/templates/fuel_quote_preview.html.tera:10]: I like the visual!

**Review by Mnguyener — COMMENTED** [review]
Good stuff, just had a question

File: src/backend/src/pricing.rs (hunk 4 above)
Comment (Brod8362) [reply src/backend/src/pricing.rs:26]: No idea, just following assignment rubric

**Pull request merged**